src/integration_tests/rpc_multiple_remote_addrs/main.cc

```diff
@@ -0,0 +1,140 @@
+#include <core/distributed.hh>
+#include <core/sleep.hh>
+#include <net/api.hh>
+// smf
+#include "integration_tests/non_root_port.h"
+#include "smf/log.h"
+#include "smf/random.h"
+#include "smf/rpc_generated.h"
+#include "smf/rpc_handle_router.h"
+#include "smf/rpc_recv_context.h"
+#include "smf/rpc_server.h"
+// templates
+#include "integration_tests/demo_service.smf.fb.h"
+
+struct client_info {
+    explicit client_info(seastar::ipv4_addr ip): client(smf_gen::demo::SmfStorageClient::make_shared(ip)), ip(ip.ip) {};
+    client_info(): client(nullptr), ip(0), ip_str("") {};
+    seastar::shared_ptr<smf_gen::demo::SmfStorageClient> client = nullptr;
+    uint32_t ip = 0;
+    char ip_str[INET_ADDRSTRLEN]{};
+};
+
+class storage_service : public smf_gen::demo::SmfStorage {
+  virtual seastar::future<smf::rpc_typed_envelope<smf_gen::demo::Response>>
+  Get(smf::rpc_recv_typed_context<smf_gen::demo::Request> &&rec) final {
+    char address_str[INET_ADDRSTRLEN] = {};
+    smf::rpc_typed_envelope<smf_gen::demo::Response> data;
+    if (rec) {
+        DLOG_DEBUG("Got request name {}", rec->name()->str());
+        inet_ntop(AF_INET, &rec.ctx->remote_address.as_posix_sockaddr_in().sin_addr, address_str, INET_ADDRSTRLEN);
+        data.data->name = seastar::to_sstring(address_str);
+    }
+    data.envelope.set_status(200);
+    return seastar::make_ready_future<
+        smf::rpc_typed_envelope<smf_gen::demo::Response>>(std::move(data));
+  }
+};
+
+#define MAX_CLIENTS 2
+
+struct ifaddrs* next_valid_interface(struct ifaddrs* ifa) {
+    while (ifa && !(ifa->ifa_addr->sa_family == AF_INET && ifa->ifa_flags & IFF_RUNNING)) {
+        ifa = ifa->ifa_next;
+    }
+    DLOG_DEBUG("Selecting interface {}", ifa->ifa_name);
+    return ifa;
+}
+
+std::array<client_info, MAX_CLIENTS>& init_clients_array(std::array<client_info, MAX_CLIENTS> &clients, struct ifaddrs *ifa, uint16_t port) {
+    for (int i = 0; i < MAX_CLIENTS; ifa = ifa->ifa_next, i++) {
+        ifa = next_valid_interface(ifa);
+        THROW_IFNULL(ifa);
+        auto sa_in = reinterpret_cast<struct sockaddr_in*>(ifa->ifa_addr);
+        auto ip = ntohl(sa_in->sin_addr.s_addr);
+        clients[i] = client_info(seastar::make_ipv4_address(ip, port));
+        THROW_IFNULL(inet_ntop(AF_INET, &sa_in->sin_addr, clients[i].ip_str, INET_ADDRSTRLEN));
+        DLOG_DEBUG("Added client ip {}", clients[i].ip_str);
+    }
+    return clients;
+}
+
+int
+main(int args, char **argv, char **env) {
+  DLOG_DEBUG("About to start the RPC test");
+  seastar::distributed<smf::rpc_server> rpc;
+
+  seastar::app_template app;
+
+  smf::random rand;
+  uint16_t random_port =
+    smf::non_root_port(rand.next() % std::numeric_limits<uint16_t>::max());
+
+  struct ifaddrs *ifaddr;
+  std::array<client_info, MAX_CLIENTS> clients;
+
+  if (getifaddrs(&ifaddr) == -1) {
+    LOG_THROW("Failed to obtain host interface list");
+  }
+
+  clients = init_clients_array(clients, ifaddr, random_port);
```

Comment:
why have 4 steps to initialize clients. 

1) define structs
2) get addrs
3) init array

i think we can use RAII (request acquisition is initialization) here. 

`
```
struct client_disapatcher {
   explicit clients(uint16_t port, uint32_t num): clietn_size(num) {
                ............ do initialization here
   }
    const uint32_t client_size;
   std::vector<client_info> clients;
}

```